### ISSUE TITLE
The multizone setting in gce.conf is ignored

### ISSUE THREAD

**Issue by jlowdermilk** [kubernetes/kubernetes#27150]
I tried spinning up a multizone cluster with 1.3.0-alpha.5, and the multizone option isn't passed through, so nodes in other zones register and are immediately deleted by node controller.

Sample output from kube-controller-manager.log
```
I0609 20:58:18.386281       9 nodecontroller.go:643] Deleting node (no longer present in cloud provider): multi-zone-minion-group-f73f
I0609 20:58:18.386318       9 nodecontroller.go:720] Recording Deleting Node multi-zone-minion-group-f73f because it's not present according to cloud provider event message for node multi-zone-minion-group-f73f
W0609 20:58:18.386359       9 nodecontroller.go:787] Missing timestamp for Node multi-zone-minion-group-sds7. Assuming now as a timestamp.
I0609 20:58:18.386538       9 event.go:216] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"multi-zone-minion-group-f73f", UID:"multi-zone-minion-group-f73f", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'DeletingNode' Node multi-zone-minion-group-f73f event: Deleting Node multi-zone-minion-group-f73f because it's not present according to cloud provider
E0609 20:58:18.462865       9 gce.go:2467] getInstanceByName/single-zone: failed to get instance multi-zone-minion-group-sds7; err: googleapi: Error 404: The resource 'projects/jeffmltest/zones/us-central1-b/instances/multi-zone-minion-group-sds7' was not found, notFound

```

It appears that we have an e2e test for this, http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce-ubernetes-lite/. However, the test has the exact same failure, and is not catching that the master is running in single-zone mode. From the latest passing run (https://console.cloud.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gce-ubernetes-lite/4643)

```
I0609 21:02:00.157122       6 nodecontroller.go:720] Recording Registered Node jenkins-e2e-minion-group-z3jn in NodeController event message for node jenkins-e2e-minion-group-z3jn
I0609 21:02:00.157140       6 nodecontroller.go:822] Node jenkins-e2e-master ReadyCondition updated. Updating timestamp.
I0609 21:02:00.157168       6 nodecontroller.go:822] Node jenkins-e2e-minion-group-kixe ReadyCondition updated. Updating timestamp.
I0609 21:02:00.157178       6 nodecontroller.go:822] Node jenkins-e2e-minion-group-okah ReadyCondition updated. Updating timestamp.
W0609 21:02:00.157185       6 nodecontroller.go:787] Missing timestamp for Node jenkins-e2e-minion-group-s3qs. Assuming now as a timestamp.
I0609 21:02:00.157537       6 event.go:216] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"jenkins-e2e-minion-group-s3qs", UID:"jenkins-e2e-minion-group-s3qs", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'RegisteredNode' Node jenkins-e2e-minion-group-s3qs event: Registered Node jenkins-e2e-minion-group-s3qs in NodeController
I0609 21:02:00.157583       6 event.go:216] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"jenkins-e2e-minion-group-u3fe", UID:"jenkins-e2e-minion-group-u3fe", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'RegisteredNode' Node jenkins-e2e-minion-group-u3fe event: Registered Node jenkins-e2e-minion-group-u3fe in NodeController
I0609 21:02:00.157597       6 event.go:216] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"jenkins-e2e-minion-group-z3jn", UID:"jenkins-e2e-minion-group-z3jn", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'RegisteredNode' Node jenkins-e2e-minion-group-z3jn event: Registered Node jenkins-e2e-minion-group-z3jn in NodeController
E0609 21:02:00.258501       6 gce.go:2492] getInstanceByName/single-zone: failed to get instance jenkins-e2e-minion-group-s3qs; err: googleapi: Error 404: The resource 'projects/k8s-jkns-e2e-gce-ubelite/zones/us-central1-a/instances/jenkins-e2e-minion-group-s3qs' was not found, notFound
I0609 21:02:00.258559       6 nodecontroller.go:643] Deleting node (no longer present in cloud provider): jenkins-e2e-minion-group-s3qs
I0609 21:02:00.258570       6 nodecontroller.go:720] Recording Deleting Node jenkins-e2e-minion-group-s3qs because it's not present according to cloud provider event message for node jenkins-e2e-minion-group-s3qs
```

Note the the gce.conf file does exist on the master, but is apparently not being plumbed through, as kube-controller-manager.log is missing the expected ["Using GCE provider config %+v"](https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/gce/gce.go#L269)

This needs to be fixed asap for 1.3. cc @quinton-hoole @andyzheng0831. kubernetes-e2e-gce-ubernetes-lite should also be failing right now, since it's supposed to test this functionality.

**Comment by andyzheng0831** [comment]
@jlowdermilk is that issue for GKE only or reproducible in oss k8s? Could you please give an instruction how to create a multi-zone cluster? I will take a look

**Comment by jlowdermilk** [comment]
It's reproducible in oss k8s, which is how I found it. Create a multi-zone cluster by following http://kubernetes.io/docs/admin/multiple-zones/#bringing-up-your-cluster.

**Comment by jlowdermilk** [comment]
As I mentioned, it's also currently reproducing in http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce-ubernetes-lite/.

**Comment by andyzheng0831** [comment]
I will take care of this issue

**Comment by jlowdermilk** [comment]
cc @chrislovecnm

**Comment by chrislovecnm** [comment]
@anndyzheng0831 thanks. I have reproduced on GCE with alpha 5. Let me know if you have questions.

**Comment by andyzheng0831** [comment]
Quick questions: 
(1) where is the file gce.conf used? I search the code but only see very limited results.

(2) Is nodecontroller running in a pod or as a part of kubelet binary?

**Comment by jlowdermilk** [comment]
(1) gce.conf is loaded (or should be loaded) at https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/gce/gce.go#L134. It's used to provide config to the gce cloudprovider plugin. See
https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/gce/gce.go#L92

(2) nodecontroller runs as part of the kube-controller-manager pod on the master.

**Comment by chrislovecnm** [comment]
I was also getting an unset var for I think KUBE_MASTER.  @jlowdermilk were you getting that as well?

**Comment by andyzheng0831** [comment]
Status update:
1. I don't think this is an issue only in GCI. I tried with all instances on ContainerVM. I hit the same issue. The second group fails to join in the cluster. "kubectl get nodes --show-labels" only shows the first group. The kube-controller-manager log also shows "Deleting node (no longer present in cloud provider): kubernetes-minion-group-8cbd"

2. I got the same error as in @chrislovecnm 's comment https://github.com/kubernetes/kubernetes/issues/27150#issuecomment-225076413, even when I use ContainerVM as master.

I will update the issue title, otherwise people does not know this is not a GCI specific issue.

**Comment by chrislovecnm** [comment]
@andyzheng0831 I am also noticing that my master is GCI (I think) an my nodes are debian ... And I selected GCI ...

**Comment by andyzheng0831** [comment]
@chrislovecnm by default you will get a cluster with master on GCI and nodes on ContainerVM. But you have a way to explicitly set what distro you want to use. For example, I just tested the pure ContainerVM-based cluster by setting the following vars. Then your cluster will has nothing to do with GCI.

export KUBE_OS_DISTRIBUTION=debian
export KUBE_GCE_MASTER_IMAGE=container-vm-v20160321

I have a theory about this issue and am trying a fix. Hopefully it can work

**Comment by chrislovecnm** [comment]
@andyzheng0831 we that documented more :) Unless you can point me that direction.  Otherwise I will filed a separate ticket.

**Comment by chrislovecnm** [comment]
@andyzheng0831 appreciate the help a TON!!!!!

**Comment by andyzheng0831** [comment]
@chrislovecnm what did you refer to "documented more"? There is a long discussion in issue #25977 about switch the default master from ContainerVM to GCI. And you can find more information about GCI in this README (https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/gci/README.md). All GCI instance config is hosted in that dir also.

**Comment by chrislovecnm** [comment]
@andyzheng0831 sorry was not clear.  Full documentation on every damn ENV var in those awesome scripts.  What they do ...  I had to pick through them by hand.  Pretty clear, but if you don't know bash ...

**Comment by andyzheng0831** [comment]
Oh, got you. You are talking about this file https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/config-default.sh. Yeah, it is not straightforward to see what is a var for. Typically, I have to grep the shell scripts in cluster/gce/ to see how to use them. So if you think this hurts, please go ahead with logging another issue

**Comment by chrislovecnm** [comment]
@andyzheng0831 yah ... you know documentation for that and other cloud providers .... lol

**Comment by andyzheng0831** [comment]
Originally I thought the issue is due to the error about unbound var KUBE_MASTER when running kube-up.sh for the second node group. After I manually export this var to the master name, the issue still happens.

@jlowdermilk as I mentioned in https://github.com/kubernetes/kubernetes/issues/27150#issuecomment-225266309, this is not a GCI issue. It is a general problem for ContainerVM also. I am not familiar with the shell scripts for cluster creation in cluster/gce. I think @zmerlynn has much better expertise.

I will assign it to @zmerlynn . You can first make it work for ContainerVM master. Setting these two vars will give you a pure ContainerVM cluster.

export KUBE_OS_DISTRIBUTION=debian
export KUBE_GCE_MASTER_IMAGE=container-vm-v20160321

**Comment by chrislovecnm** [comment]
@zmerlynn to be clear this is a huge blocker for me ... I need it fixed ASAP, and preferably in a release.  We are trying to get multiple zone and multiple DC Cassandra working.

cc: @bgrant0607 @thockin

**Comment by chrislovecnm** [comment]
cc: @paralin we cannot test racks until this is fixed.  We can do multiple DC, but racks in the same DC is currently broken in 1.3 alpha 5.

**Comment by davidopp** [comment]
cc/ @quinton-hoole

**Comment by chrislovecnm** [comment]
@davidopp much thanks

**Comment by bgrant0607** [comment]
@bprashanth @erictune

**Comment by zmerlynn** [comment]
I'm not clear I have the bandwidth to work on this right now. Reassigning for triage to the Ubernetes team.

**Comment by chrislovecnm** [comment]
@bgrant0607 we go anyone available to look at this?

**Comment by bprashanth** [comment]
I'm going to guess we're not actually mounting gce.conf in the container. We should have something like: https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/l7-gcp/glbc.manifest#L51 which should be coming in through https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/kube-controller-manager/kube-controller-manager.manifest#L50. If that is happening, the config might just be wrong.

**Comment by jlowdermilk** [comment]
I just successfully created a multi-zone cluster with a CVM master using k8s from HEAD. So I think my original statement that this was broken for GCI still stands.
```
kubectl get nodes --show-labels
NAME                           STATUS                     AGE       LABELS
multi-zone-master              Ready,SchedulingDisabled   9m        beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=n1-standard-4,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-central1,failure-domain.beta.kubernetes.io/zone=us-central1-f,kubernetes.io/hostname=multi-zone-master
multi-zone-minion-group-0qfb   Ready                      8m        beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=n1-standard-1,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-central1,failure-domain.beta.kubernetes.io/zone=us-central1-f,kubernetes.io/hostname=multi-zone-minion-group-0qfb
multi-zone-minion-group-71vc   Ready                      3m        beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=n1-standard-1,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-central1,failure-domain.beta.kubernetes.io/zone=us-central1-b,kubernetes.io/hostname=multi-zone-minion-group-71vc
multi-zone-minion-group-eyro   Ready                      3m        beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=n1-standard-1,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-central1,failure-domain.beta.kubernetes.io/zone=us-central1-b,kubernetes.io/hostname=multi-zone-minion-group-eyro
multi-zone-minion-group-ohzc   Ready                      8m        beta.kubernetes.io/arch=amd64,beta.kubernetes.io/instance-type=n1-standard-1,beta.kubernetes.io/os=linux,failure-domain.beta.kubernetes.io/region=us-central1,failure-domain.beta.kubernetes.io/zone=us-central1-f,kubernetes.io/hostname=multi-zone-minion-group-ohzc
```

**Comment by andyzheng0831** [comment]
Can you list the detailed steps about your reproduction. I tried it before. The second group joined in but after a while it was removed by the master. Let me use your recipe to try again

**Comment by jlowdermilk** [comment]
Steps below. I built from head (`v1.3.0-alpha.5.446+0d02f8c0f6c0f5`)

```
make release-skip-tests
...
export KUBE_GCE_INSTANCE_PREFIX=multi-zone
export KUBE_GCE_MASTER_IMAGE=container-vm-v20160321
export KUBE_OS_DISTRIBUTION=debian
export NUM_NODES=2
export MASTER_SIZE=n1-standard-4
export NODE_SIZE=n1-standard-1

# create master and first zone nodes
MULTIZONE=true KUBERNETES_PROVIDER=gce KUBE_GCE_ZONE=us-central1-f cluster/kube-up.sh
# create second zone nodes
KUBE_USE_EXISTING_MASTER=true MULTIZONE=true KUBERNETES_PROVIDER=gce KUBE_GCE_ZONE=us-central1-b cluster/kube-up.sh
```
Then did `kubectl get nodes --watch` until 2 new nodes joined. Output I pasted [above](https://github.com/kubernetes/kubernetes/issues/27150#issuecomment-225712166) was after the new nodes became Ready.

**Comment by jlowdermilk** [comment]
On cvm, the config appears to get plumbed through as expected. From `/var/log/kube-controller-manager.log`:
```
I0613 21:08:51.854465       5 gce.go:269] Using GCE provider config {Global:{TokenURL: TokenBody: ProjectID: Networ
kName: NodeTags:[multi-zone-minion] Multizone:true}}
```
I have a gci cluster now with the same build, and one node each in us-central1-f, us-central1-b. It's missing the above log indicating config was loaded, and is deleting the us-central1-b node because it's in single-zone mode:
```
E0613 21:40:57.387182       9 gce.go:2506] getInstanceByName/single-zone: failed to get instance multi-zone-gci-min
ion-group-25su; err: googleapi: Error 404: The resource 'projects/jeffmltest/zones/us-central1-f/instances/multi-zo
ne-gci-minion-group-25su' was not found, notFound
I0613 21:40:57.387260       9 nodecontroller.go:643] Deleting node (no longer present in cloud provider): multi-zon
e-gci-minion-group-25su
```

**Comment by jlowdermilk** [comment]
@bprashanth's suggestion is correct, gce.conf isn't showing up as a volume/mount in the manifest.
```
jeffml@multi-zone-gci-master ~ $ sudo cat /etc/kubernetes/manifests/kube-controller-manager.manifest              













{
"apiVersion": "v1",
"kind": "Pod",
"metadata": {
  "name":"kube-controller-manager",
  "namespace": "kube-system",
  "labels": {
    "tier": "control-plane",
    "component": "kube-controller-manager"
  }
},
"spec":{
"hostNetwork": true,
"containers":[
    {
    "name": "kube-controller-manager",
    "image": "gcr.io/google_containers/kube-controller-manager:87205ee0727c88dab5e44fee708ab444",
    "resources": {
      "requests": {
        "cpu": "200m"
      }
    },
    "command": [
                 "/bin/sh",
                 "-c",
                 "/usr/local/bin/kube-controller-manager --v=2  --cloud-provider=gce --master=127.0.0.1:8080 --root-ca-file=/etc/srv/kubernetes/ca.crt --service-account-private-key-file=/etc/srv/kubernetes/server.key --cluster-name=multi-zone-gci --cluster-cidr=10.244.0.0/14 --service-cluster-ip-range=10.0.0.0/16 --allocate-node-cidrs=true 1>>/var/log/kube-controller-manager.log 2>&1"
               ],
    "livenessProbe": {
      "httpGet": {
        "host": "127.0.0.1",
        "port": 10252,
        "path": "/healthz"
      },
      "initialDelaySeconds": 15,
      "timeoutSeconds": 15
    },
    "volumeMounts": [
        
        
        { "name": "srvkube",
        "mountPath": "/etc/srv/kubernetes",
        "readOnly": true},
        { "name": "logfile",
        "mountPath": "/var/log/kube-controller-manager.log",
        "readOnly": false},
        { "name": "etcssl",
        "mountPath": "/etc/ssl",
        "readOnly": true},
        { "name": "varssl",
        "mountPath": "/var/ssl",
        "readOnly": true},
        { "name": "etcopenssl",
        "mountPath": "/etc/openssl",
        "readOnly": true},
        { "name": "etcpki",
        "mountPath": "/etc/pki",
        "readOnly": true}
      ]
    }
],
"volumes":[
  
  
  { "name": "srvkube",
    "hostPath": {
        "path": "/etc/srv/kubernetes"}
  },
  { "name": "logfile",
    "hostPath": {
        "path": "/var/log/kube-controller-manager.log"}
  },
  { "name": "etcssl",
    "hostPath": {
        "path": "/etc/ssl"}
  },
  { "name": "varssl",
    "hostPath": {
        "path": "/var/ssl"}
  },
  { "name": "etcopenssl",
    "hostPath": {
        "path": "/etc/openssl"}
  },
  { "name": "etcpki",
    "hostPath": {
        "path": "/etc/pki"}
  }
]
}}
```

**Comment by andyzheng0831** [comment]
I can see there is a bug in the GCI config, which missed the mounted volume. I made a fix and trying it.

**Comment by andyzheng0831** [comment]
A question (may be stupid): two node groups in us-central1-f and us-central1-b can form a multi-zone cluster. But groups in us-central1-f and us-east1-b cannot?

I think this is why I cannot create a multi-zone cluster using ContainerVM master. I set two zones to us-central1-f and us-east1-b

**Comment by jlowdermilk** [comment]
@andyzheng0831, that would be multi-region. I don't think there's a functional reason we can't do it, but it wouldn't surprise me if controller manager doesn't allow cross-region, since it would incur network charges and have higher latency.

**Comment by roberthbailey** [comment]
@andyzheng0831 all nodes for the cluster must be in the same region.

**Comment by andyzheng0831** [comment]
The fix https://github.com/kubernetes/kubernetes/pull/27313 is sent out for review. The root cause of this bug: initially we only had one case of using /etc/gce.conf, and then more cases were added. But we did not update the condition at https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/gci/configure-helper.sh#l476. As a result, the new usage cases of gce.conf, e.g., multizone, were ignored.

**Comment by andyzheng0831** [comment]
The fix was merged into master branch and will be cherrypicked into release-1.3. Close this issue

**Comment by jlowdermilk** [comment]
Reopening to track updating the e2e test suite that should have caught this.

**Comment by chrislovecnm** [comment]
Anyone know when the next alpha or beta binary will be cut that will catch this?

**Comment by jlowdermilk** [comment]
I believe we're also waiting on a fix for #27197, but should be soon.

**Comment by jlowdermilk** [comment]
Opened #27372 to track updating tests to catch this issue. Closing again to not confuse other trackers.

**Comment by roberthbailey** [comment]
What version of the autoscaler are you running (older versions didn't scale evenly across multiple zones)?

@mwielgus

**Comment by dineshsckloud9** [comment]
i'm running cluster-autoscaler     v1.1.0, running on GCE kubernetes 1.9.2

**Comment by dineshsckloud9** [comment]
How to find the exact version? could you please help me?

I sent you the docker image version

**Comment by roberthbailey** [comment]
The docker image version should be sufficient. @mwielgus - can you look at this when you get a chance?

**Comment by dineshsckloud9** [comment]
Thanks Robert... Version is fine,

Facing new problem,


Multizone scaling with MIG per zone seems working but could not launch more
than 3 instances.. MIG showing 1 -> 2 but its not launching in gce... What
could be the problem?

I0127 04:52:37.565409 5 scale_up.go:110] Skipping node group
https://content.googleapis.com/compute/v1/projects/global-snow-193212/zones/europe-west1-b/instanceGroups/kubernetes-minion-group
-
max size reached
W0127 04:52:37.566044 5 scale_up.go:99] Node group
https://content.googleapis.com/compute/v1/projects/global-snow-193212/zones/europe-west1-d/instanceGroups/kubernetes-minion-group
is
not ready for scaleup
W0127 04:52:37.566348 5 scale_up.go:99] Node group
https://content.googleapis.com/compute/v1/projects/global-snow-193212/zones/europe-west1-c/instanceGroups/kubernetes-minion-group
is
not ready for scaleup

Why it is not ready?

**Comment by roberthbailey** [comment]
> MIG showing 1 -> 2 but its not launching in gce

This is a sign that the MIG is attempting to add a VM but is unable to. It usually will give an indication of why (e.g. quota issue, etc).

**Comment by dineshsckloud9** [comment]
This is fixed and auto scaling is working for GCE kubernetes...

Will the same thing work for openshift?

Here, my openshift cluster should be in GCE, once I create it, I will try
to run Custer autoscaler pod with node-group and all in it, will it
scale-up nodes?
